<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: (Prediction Using Models Informed by Chromatin conformations and
Epigenomics</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> (Prediction Using Models Informed by Chromatin conformations and
Epigenomics
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;PUMICE&rsquo; version 0.0.0.9000</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="calc_R2.html">calc_R2</a></td>
<td>Function for calculating coefficient of determination#' Read plink file through R</td></tr>
<tr><td style="width: 25%;"><a href="constant_process.html">constant_process</a></td>
<td>Function for mapping snps to genes in 3D method (e.g. Domain, Loop, TAD)</td></tr>
<tr><td style="width: 25%;"><a href="elnet.html">elnet</a></td>
<td>Cross-validation</td></tr>
<tr><td style="width: 25%;"><a href="elnet.cv.html">elnet.cv</a></td>
<td>Nested Cross-validation</td></tr>
<tr><td style="width: 25%;"><a href="expression_scale.html">expression_scale</a></td>
<td>Function for scaling expression dataframe to mean of zero</td></tr>
<tr><td style="width: 25%;"><a href="file_helper.html">file_helper</a></td>
<td>Function for splitting jobs</td></tr>
<tr><td style="width: 25%;"><a href="func_process.html">func_process</a></td>
<td>Function for mapping snps to genes in 3D method (e.g. Domain, Loop, TAD)</td></tr>
<tr><td style="width: 25%;"><a href="geno_coord.html">geno_coord</a></td>
<td>Function for filling in snp coordinates</td></tr>
<tr><td style="width: 25%;"><a href="geno_fill.html">geno_fill</a></td>
<td>Function for filling in missing genotype</td></tr>
<tr><td style="width: 25%;"><a href="geno_scale.html">geno_scale</a></td>
<td>Function for scaling genotype dataframe to mean of zero and sd of one</td></tr>
<tr><td style="width: 25%;"><a href="pchic_process.html">pchic_process</a></td>
<td>Function for mapping snps to genes in 3D method pchic</td></tr>
<tr><td style="width: 25%;"><a href="PUMICE.compute_weight.html">PUMICE.compute_weight</a></td>
<td>Create gene expression prediction model using window type and penalty factor derived from the CV step</td></tr>
<tr><td style="width: 25%;"><a href="PUMICE.nested_cv.html">PUMICE.nested_cv</a></td>
<td>Run nested cross-validation to determine which window type and penalty factor are optimal</td></tr>
<tr><td style="width: 25%;"><a href="read_plink_custom.html">read_plink_custom</a></td>
<td>Read plink file through R</td></tr>
</table>
</body></html>
